packages/modules/internal_openwb/socket.py

```diff
@@ -62,23 +61,32 @@ def get_values(self) -> Tuple[ChargepointState, float]:
         return self.chargepoint_state, self.set_current_evse
 
     def __open_actor(self):
-        GPIO.output(23, GPIO.LOW)
-        GPIO.output(26, GPIO.HIGH)
-        time.sleep(2)
-        GPIO.output(26, GPIO.LOW)
-        log.debug("Aktor auf")
-        self.actor_moves += 1
+        self.__set_actor(open=True)
 
     def __close_actor(self):
-        GPIO.output(23, GPIO.HIGH)
+        self.__set_actor(open=False)
+
+    def __set_actor(self, open: bool):
+        GPIO.output(23, GPIO.LOW if open else GPIO.HIGH)
         GPIO.output(26, GPIO.HIGH)
-        time.sleep(3)
+        time.sleep(2 if open else 3)
         GPIO.output(26, GPIO.LOW)
-        log.debug("Aktor zu")
-        self.actor_moves += 1
-
-    def cooldown_neccessary(self) -> bool:
-        return self.actor_moves >= 10
+        log.debug("Actor opened" if open else "Actor closed")
+        self.cooldown_tracker.move()
 
     def perform_actor_cooldown(self):
         time.sleep(300)
+
+
+class CooldownTracker:
```

Comment:
Ich habe gerade nochmal darüber nachgedacht, ob es sich noch weiter vereinfachen lässt. Mir ist dies hier eingefallen:
```python
class RateLimiter:
    def __init__(self, max_calls: int, max_seconds: int):
        self.movement_times = [0]*max_calls
        self.max_seconds = max_seconds
        self.counter = 0

    def __call__(self, delegate: Callable):
        @functools.wraps(delegate)
        def wrapper(*args, **kwargs):
            now = time.time()
            sleep_needed = self.max_seconds - (now - self.movement_times[self.counter])
            if sleep_needed > 0:
                time.sleep(sleep_needed)
            self.movement_times[self.counter] = time.time()
            self.counter = (self.counter + 1) % len(self.movement_times)
            return delegate(*args, **kwargs)
        return wrapper
```
Das könnte man dann einfach als decorator verwenden:
```python
@RateLimiter(max_calls=10, max_seconds=300)
def __set_actor(self, open: bool):
    # ...
```
Das würde dazu führen, dass der RateLimiter alles managt. Wenn mehr als 10 Aufrufe innerhalb der letzten 300 Sekunden waren, dann schläft er solange, bis das Fenster wieder eingehalten wird. Das wäre dann noch etwas sparsamer als die bisherige Variante. So wie es jetzt ist wäre es möglich, dass es 9 Aufrufe kurz hintereinander gibt und dann 295 Sekunden den 10. Aufruf und dann würde er sich sich gleich volle 5 Minuten schlafen legen. Mit der obigen Implementierung würde er nur 5 Sekunden schlafen, weil das dafür, dass das Fenster eingehalten wird reicht. Allerdings kann man dadurch nicht mehr steuern wann der Schlaf durchgeführt wird. Ob das nun ein Problem ist dürfte von weiteren Faktoren abhängen, die ich nicht Überblicke. Wenn es genügt einfach nur das Fenster einzuhalten und das sleep nicht ausgerechnet an der Stelle stört, dann wäre der decorator eine schönere Lösung.

Hier noch Tests dazu:
```python
class FakeClock:
    def __init__(self):
        self.time = 1000.0

    def sleep(self, seconds: float):
        self.time += seconds


@pytest.fixture
def fake_clock(monkeypatch):
    fake_clock = FakeClock()
    monkeypatch.setattr(time, 'time', lambda: fake_clock.time)
    monkeypatch.setattr(time, 'sleep', lambda seconds: fake_clock.sleep(seconds))
    return fake_clock


def test_rate_limiter_calls_delegate():
    # setup
    mock = Mock(return_value="some return value")
    rate_limited_mock = RateLimiter(max_calls=2, max_seconds=10)(mock)

    # execution
    result = rate_limited_mock(1, 2, a="x", b="y")

    # evaluation
    assert result == "some return value"
    assert mock.call_count == 1
    mock.assert_called_with(1, 2, a="x", b="y")


def test_rate_limiter_sleeps_as_needed(fake_clock: FakeClock):
    # setup
    mock = Mock()
    rate_limited_mock = RateLimiter(max_calls=2, max_seconds=10)(mock)

    # execution
    # 1. We can make two calls without any limitation
    rate_limited_mock(1)
    rate_limited_mock(2)
    assert fake_clock.time == 1000.0
    # The next call is only possible after 10s cooldown
    rate_limited_mock(3)
    assert fake_clock.time == 1010.0
    # However for the forth call no cooldown is needed
    rate_limited_mock(4)
    assert fake_clock.time == 1010.0
    # Now we need a cooldown again, but now we simulate that some time passes "naturally" and thus no sleep is needed:
    fake_clock.sleep(10)
    rate_limited_mock(5)
    rate_limited_mock(6)
    assert fake_clock.time == 1020.0
    # Also if only a partial time passes "naturally", only the window needs to be completed:
    fake_clock.sleep(7)
    rate_limited_mock(7)
    assert fake_clock.time == 1030.0
```

Reply:
Die Idee ist gut, allerdings würde dann während dem Cooldown der gesamte Daemon pausieren. Das war bisher so, soll aber verbessert werden.
Ich habe es so abgeändert, dass wenn ein Cooldown erforderlich ist, die Wrapperfunktion nicht aufgerufen wird. Es findet dann in der Zeit keine Bewegung des Aktors statt.